File: Problem/By section/Binary Search/Problem 540/problem_540.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Time: O(log(n)), Space: O(1)
    int singleNonDuplicate(vector<int>& nums) {
        int size = nums.size();
        int left = 0, right = size - 1;
        int mid;

        while(left < right) {
            mid = left + (right - left) / 2;
            // if we reach left or right bound
            if(mid == 0 || mid == size - 1) {
                break;
            }else if(nums[mid - 1] != nums[mid] && nums[mid + 1] != nums[mid]) {
                return nums[mid];
            }else if(nums[mid - 1] != nums[mid]) { // if the mid is not equal to the left
                // and the num of number left in the left part is not even 
                if((mid - left) % 2 != 0) {
                    // the answer must be in the left part
                    right = mid - 1;
                }else {
                    left = mid;
                }
            }else if(nums[mid + 1] != nums[mid]) { // same for rights
                if((right - mid) % 2 != 0) {
                    left = mid + 1;
                }else {
                    right = mid;
                }
            }
        }

        return nums[left];
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> input = {1,1,2,2,3,3,4,8,8};
    cout << s.singleNonDuplicate(input) << endl;
    return 0;
}

File: Problem/By section/Binary Search/Problem 540/problem_540.py
class Solution(object):
    # Time: O(log(n)), Space: O(1)
    def singleNonDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        n = len(nums)

        if n == 1:
            return nums[0]

        left, right = 0, n

        while left < right:
            mid = left + (right - left) // 2

            if self.is_target_number(nums, mid):
                return nums[mid]
            else:
                if nums[mid] == nums[mid + 1]:
                    if (right - mid) % 2 == 0:
                        right = mid
                    else:
                        left = mid + 2
                elif nums[mid] == nums[mid - 1]:
                    if (mid - 2 - left + 1) % 2 == 0:
                        left = mid + 1
                    else:
                        right = mid - 1
                else:
                    assert False
        

        return nums[left]



    def is_target_number(self, nums, idx):
        if idx == 0 and nums[idx] != nums[idx + 1]:
            return True
        
        if (idx == len(nums) - 1) and nums[idx] != nums[idx - 1]:
            return True

        return nums[idx] != nums[idx - 1] and nums[idx] != nums[idx + 1]

if __name__ == "__main__":
    solution = Solution()
    print(solution.singleNonDuplicate([3,3,7,7,11,11,12]))